File: backend/subtitle_wrap.py
"""Wrap Chinese subtitle text to multi-line display.

Algorithm (ZH path):
  1. If text length <= cap + tail_tolerance -> single line
  2. Otherwise scan [1, cap] for break points by priority:
     - HARD (。！？!?) score 100
     - SOFT (，、；：,;:) score 50
     - PAREN_CLOSE (）」』]) score 30
     - PAREN_OPEN_LOOKAHEAD (next char in （「『() score 25
     - tiebreaker: prefer higher index (longer first chunk)
  3. If no break point found -> hard cut at cap, flag hard_cut=True
  4. Last line allows cap + tail_tolerance to absorb trailing punctuation
  5. After max_lines reached, append leftover to last line (avoid data loss)

Algorithm (EN path — smart-break v2):
  Score-based break with cap-aware remaining-content lookahead. Prefers
  sentence/clause punctuation over connectors over prepositions over plain
  whitespace. Penalises breaks that split a Title-case word pair (proper
  nouns like "David Alaba"). Falls back to latest fitting position when no
  break candidate leaves enough room for remaining content.
"""
from dataclasses import dataclass, field
from typing import List
import re

HARD_BREAKS = "。！？!?"
SOFT_BREAKS = "，、；：,;:"
PAREN_CLOSE = "）」』)]"
PAREN_OPEN = "（「『(["

_EN_HARD = set(".!?")
_EN_SOFT = set(",;:")
_EN_CONNECTORS = {
    "and", "but", "or", "nor", "so", "yet", "when", "after", "before",
    "while", "because", "although", "since", "though", "if", "unless",
    "until", "as",
}
_EN_PREPOSITIONS = {
    "to", "of", "in", "on", "at", "with", "for", "from", "by", "into",
    "onto", "upon", "about", "between", "through", "over", "under", "against",
}


@dataclass
class WrapResult:
    lines: List[str] = field(default_factory=list)
    hard_cut: bool = False


def _find_break(remaining: str, cap: int, tail_tolerance: int = 0) -> int:
    """Return the best break index. Searches [1, cap], then [cap+1, cap+tail_tolerance].

    Returns -1 if no break point found in either range.
    """
    best = -1
    best_score = -1
    primary_limit = min(cap, len(remaining))
    extended_limit = min(cap + tail_tolerance, len(remaining))

    # Pass 1: primary range [1, cap]
    for i in range(1, primary_limit + 1):
        ch = remaining[i - 1]
        score = 0
        if ch in HARD_BREAKS:
            score = 100
        elif ch in SOFT_BREAKS:
            score = 50
        elif ch in PAREN_CLOSE:
            score = 30
        elif i < len(remaining) and remaining[i] in PAREN_OPEN:
            score = 25
        if score > 0:
            score += i  # tiebreaker: prefer longer first chunk
            if score > best_score:
                best_score = score
                best = i

    if best != -1:
        return best

    # Pass 2: extended range [cap+1, cap+tail_tolerance], only HARD/SOFT (no paren tiebreaks)
    for i in range(primary_limit + 1, extended_limit + 1):
        ch = remaining[i - 1]
        if ch in HARD_BREAKS or ch in SOFT_BREAKS:
            return i  # first match in extended range -- short-circuit

    return -1


def wrap_zh(text: str, cap: int = 23, max_lines: int = 3, tail_tolerance: int = 3) -> WrapResult:
    # V_R11 Bug #1: defensive clamp — prevent silent text drop on bad config
    cap = max(1, cap or 1)
    max_lines = max(1, max_lines or 1)
    tail_tolerance = max(0, tail_tolerance or 0)
    text = (text or "").strip()
    if not text:
        return WrapResult(lines=[], hard_cut=False)
    if len(text) <= cap + tail_tolerance:
        return WrapResult(lines=[text], hard_cut=False)

    lines: List[str] = []
    remaining = text
    hard_cut = False

    while remaining and len(lines) < max_lines:
        if len(remaining) <= cap + tail_tolerance:
            lines.append(remaining)
            remaining = ""
            break
        best = _find_break(remaining, cap, tail_tolerance)
        if best == -1:
            best = cap
            hard_cut = True
        lines.append(remaining[:best].rstrip())
        remaining = remaining[best:].lstrip()

    if remaining and lines:
        # max_lines reached but content remains -- append to last line (no data loss)
        lines[-1] = lines[-1] + remaining

    return WrapResult(lines=lines, hard_cut=hard_cut)


# === Option D additions ===
_HAS_ZH = re.compile(r'[一-鿿　-〿＀-￯]')


def _is_zh_text(text: str) -> bool:
    return bool(_HAS_ZH.search(text or ""))


def _is_titlecase_word(word: str) -> bool:
    """Word starts with capital and rest is lowercase (handles trailing punct)."""
    stripped = re.sub(r"[^\w]", "", word)
    if not stripped:
        return False
    if len(stripped) == 1:
        return stripped[0].isupper()
    return stripped[0].isupper() and stripped[1:].islower()


def _detect_titlecase_pairs(words: List[str]) -> set:
    """Indices i where words[i-1] and words[i] form a multi-word proper noun.

    Skip pairs straddling sentence boundaries (prev word ends in HARD punct).
    Returns the set of "second-word" indices — breaking BEFORE such an index
    splits the proper-noun pair.
    """
    pairs = set()
    for i in range(1, len(words)):
        prev = words[i - 1]
        if prev and prev[-1] in _EN_HARD:
            continue
        if _is_titlecase_word(prev) and _is_titlecase_word(words[i]):
            pairs.add(i)
    return pairs


def _remaining_fits(words: List[str], start: int, lines_left: int,
                    cap: int, tail_tolerance: int) -> bool:
    """Greedy-fit check: can words[start:] fit in lines_left lines of budget cap+tail?"""
    if start >= len(words):
        return True
    if lines_left <= 0:
        return False
    budget = cap + tail_tolerance
    used = 0
    cur = 0
    for w in words[start:]:
        wl = len(w)
        if cur == 0:
            cur = wl
        elif cur + 1 + wl <= budget:
            cur += 1 + wl
        else:
            used += 1
            cur = wl
            if used >= lines_left:
                return False
    return used + (1 if cur > 0 else 0) <= lines_left


def _wrap_en(text: str, cap: int, max_lines: int, tail_tolerance: int) -> WrapResult:
    """Smart-break EN wrap: punct/conn/prep priority + Title-case-pair avoidance.

    All words preserved (last line absorbs leftovers).
    """
    # V_R11 Bug #1: defensive clamp — prevent silent text drop on bad config
    cap = max(1, cap or 1)
    max_lines = max(1, max_lines or 1)
    tail_tolerance = max(0, tail_tolerance or 0)
    text = (text or "").strip()
    if not text:
        return WrapResult(lines=[], hard_cut=False)
    if len(text) <= cap + tail_tolerance:
        return WrapResult(lines=[text], hard_cut=False)

    words = text.split()
    locked_pairs = _detect_titlecase_pairs(words)
    lines: List[str] = []
    i = 0
    hard_cut = False

    while i < len(words) and len(lines) < max_lines:
        # Last allowed line: gobble all remaining words (data-loss prevention)
        if len(lines) == max_lines - 1:
            lines.append(" ".join(words[i:]))
            i = len(words)
            break

        best_j = -1
        best_score = -float("inf")
        cur_len = 0
        latest_fitting_j = i + 1

        for j in range(i, len(words)):
            wl = len(words[j])
            new_len = cur_len + (1 if j > i else 0) + wl
            if new_len > cap + tail_tolerance:
                # Past hard limit; if even first word exceeds, force one-word line
                if j == i:
                    cur_len = wl
                    latest_fitting_j = i + 1
                break
            cur_len = new_len
            latest_fitting_j = j + 1

            if j + 1 >= len(words):
                continue  # nothing follows — no break to score

            distance = abs(cur_len - cap)
            score = 10  # baseline whitespace
            last_ch = words[j][-1] if words[j] else ""
            if last_ch in _EN_HARD:
                score = 100
            elif last_ch in _EN_SOFT:
                score = 70

            nxt_clean = re.sub(r"[^\w]", "", words[j + 1]).lower()
            if nxt_clean in _EN_CONNECTORS:
                score = max(score, 50)
            elif nxt_clean in _EN_PREPOSITIONS:
                # v3: penalise breaks that strand a preposition at line start
                # (Netflix style: keep prepositional phrase intact)
                score -= 40

            # Penalise breaks that split a Title-case proper-noun pair
            if (j + 1) in locked_pairs:
                score -= 80

            score -= distance * 2

            # Lookahead: only accept break if remainder fits in remaining lines
            lines_left = max_lines - len(lines) - 1
            if not _remaining_fits(words, j + 1, lines_left, cap, tail_tolerance):
                continue

            if score > best_score:
                best_score = score
                best_j = j + 1

        if best_j <= i:
            # No lookahead-safe candidate — fall back to greedy (latest fitting)
            best_j = latest_fitting_j
            lines_left = max_lines - len(lines) - 1
            if not _remaining_fits(words, best_j, lines_left, cap, tail_tolerance):
                hard_cut = True

        lines.append(" ".join(words[i:best_j]))
        i = best_j

    if i < len(words) and lines:
        # Hit max_lines with content remaining — append to last line
        lines[-1] = lines[-1] + " " + " ".join(words[i:])

    if any(len(l) > cap + tail_tolerance for l in lines):
        hard_cut = True

    return WrapResult(lines=lines, hard_cut=hard_cut)


PRESETS = {
    "netflix_originals": {
        "zh": {"line_cap": 16, "max_lines": 2, "tail_tolerance": 2},
        "en": {"line_cap": 42, "max_lines": 2, "tail_tolerance": 4},
    },
    "netflix_general": {
        "zh": {"line_cap": 23, "max_lines": 2, "tail_tolerance": 3},
        "en": {"line_cap": 42, "max_lines": 2, "tail_tolerance": 4},
    },
    "broadcast": {
        "zh": {"line_cap": 28, "max_lines": 3, "tail_tolerance": 3},
        "en": {"line_cap": 50, "max_lines": 3, "tail_tolerance": 5},
    },
}
DEFAULT_PRESET = "broadcast"


def resolve_wrap_config(font_config: dict) -> dict:
    """Resolve final wrap config from font_config.

    Returns: {"enabled": bool, "zh": {line_cap, max_lines, tail_tolerance}, "en": {...}}
    Resolution order:
      1. font_config["line_wrap"] explicit fields override (apply to BOTH zh + en)
      2. font_config["subtitle_standard"] preset
      3. DEFAULT_PRESET ("broadcast")
    Explicit `line_wrap` overrides apply to BOTH zh and en (legacy single-cap compat).
    """
    standard = font_config.get("subtitle_standard")
    base_preset = PRESETS.get(standard, PRESETS[DEFAULT_PRESET])
    zh_cfg = dict(base_preset["zh"])
    en_cfg = dict(base_preset["en"])

    explicit = font_config.get("line_wrap") or {}
    enabled = explicit.get("enabled", True)
    # Explicit overrides apply to BOTH zh and en (legacy compat — old single-cap config)
    for key in ("line_cap", "max_lines", "tail_tolerance"):
        if key in explicit:
            zh_cfg[key] = explicit[key]
            en_cfg[key] = explicit[key]
    return {"enabled": enabled, "zh": zh_cfg, "en": en_cfg}


def wrap_with_config(text: str, font_config: dict) -> WrapResult:
    """Apply wrap based on resolved config. Detects EN vs ZH by character set."""
    cfg = resolve_wrap_config(font_config)
    if not cfg["enabled"]:
        text = (text or "").strip()
        return WrapResult(lines=[text] if text else [], hard_cut=False)
    sub = cfg["zh"] if _is_zh_text(text) else cfg["en"]
    if _is_zh_text(text):
        return wrap_zh(text, cap=sub["line_cap"], max_lines=sub["max_lines"], tail_tolerance=sub["tail_tolerance"])
    return _wrap_en(text, cap=sub["line_cap"], max_lines=sub["max_lines"], tail_tolerance=sub["tail_tolerance"])

File: frontend/js/subtitle-wrap.js
// frontend/js/subtitle-wrap.js
// Subtitle line-wrap algorithm. Mirror of backend/subtitle_wrap.py (Option D).
// Exposes window.SubtitleWrap.wrapZh + wrapEn + wrapWithConfig + resolveWrapConfig + isZhText + PRESETS

(function () {
  "use strict";

  const HARD_BREAKS = "。！？!?";
  const SOFT_BREAKS = "，、；：,;:";
  const PAREN_CLOSE = "）」』)]";
  const PAREN_OPEN = "（「『([";

  const EN_HARD = ".!?";
  const EN_SOFT = ",;:";
  const EN_CONNECTORS = new Set([
    "and", "but", "or", "nor", "so", "yet", "when", "after", "before",
    "while", "because", "although", "since", "though", "if", "unless",
    "until", "as",
  ]);
  const EN_PREPOSITIONS = new Set([
    "to", "of", "in", "on", "at", "with", "for", "from", "by", "into",
    "onto", "upon", "about", "between", "through", "over", "under", "against",
  ]);

  const HAS_ZH = /[一-鿿　-〿＀-￯]/;

  function isZhText(text) {
    return HAS_ZH.test(text || "");
  }

  const PRESETS = {
    netflix_originals: {
      zh: { line_cap: 16, max_lines: 2, tail_tolerance: 2 },
      en: { line_cap: 42, max_lines: 2, tail_tolerance: 4 },
    },
    netflix_general: {
      zh: { line_cap: 23, max_lines: 2, tail_tolerance: 3 },
      en: { line_cap: 42, max_lines: 2, tail_tolerance: 4 },
    },
    broadcast: {
      zh: { line_cap: 28, max_lines: 3, tail_tolerance: 3 },
      en: { line_cap: 50, max_lines: 3, tail_tolerance: 5 },
    },
  };
  const DEFAULT_PRESET = "broadcast";

  function findBreak(remaining, cap, tailTolerance) {
    let best = -1;
    let bestScore = -1;
    const primaryLimit = Math.min(cap, remaining.length);
    const extendedLimit = Math.min(cap + tailTolerance, remaining.length);

    for (let i = 1; i <= primaryLimit; i++) {
      const ch = remaining[i - 1];
      let score = 0;
      if (HARD_BREAKS.includes(ch)) score = 100;
      else if (SOFT_BREAKS.includes(ch)) score = 50;
      else if (PAREN_CLOSE.includes(ch)) score = 30;
      else if (i < remaining.length && PAREN_OPEN.includes(remaining[i])) score = 25;
      if (score > 0) {
        score += i;
        if (score > bestScore) {
          bestScore = score;
          best = i;
        }
      }
    }
    if (best !== -1) return best;

    for (let i = primaryLimit + 1; i <= extendedLimit; i++) {
      const ch = remaining[i - 1];
      if (HARD_BREAKS.includes(ch) || SOFT_BREAKS.includes(ch)) return i;
    }
    return -1;
  }

  function wrapZh(text, options) {
    const cap = (options && options.cap) || 23;
    const maxLines = (options && options.maxLines) || 3;
    const tailTolerance = (options && options.tailTolerance != null) ? options.tailTolerance : 3;

    const trimmed = (text || "").trim();
    if (!trimmed) return { lines: [], hardCut: false };
    if (trimmed.length <= cap + tailTolerance) return { lines: [trimmed], hardCut: false };

    const lines = [];
    let remaining = trimmed;
    let hardCut = false;

    while (remaining && lines.length < maxLines) {
      if (remaining.length <= cap + tailTolerance) {
        lines.push(remaining);
        remaining = "";
        break;
      }
      let best = findBreak(remaining, cap, tailTolerance);
      if (best === -1) {
        best = cap;
        hardCut = true;
      }
      lines.push(remaining.slice(0, best).replace(/\s+$/, ""));
      remaining = remaining.slice(best).replace(/^\s+/, "");
    }

    if (remaining && lines.length > 0) {
      lines[lines.length - 1] = lines[lines.length - 1] + remaining;
    }

    return { lines, hardCut };
  }

  function isTitlecaseWord(word) {
    const stripped = (word || "").replace(/[^\w]/g, "");
    if (!stripped) return false;
    if (stripped.length === 1) return /[A-Z]/.test(stripped[0]);
    return /[A-Z]/.test(stripped[0]) && stripped.slice(1) === stripped.slice(1).toLowerCase();
  }

  function detectTitlecasePairs(words) {
    const pairs = new Set();
    for (let i = 1; i < words.length; i++) {
      const prev = words[i - 1];
      if (prev && EN_HARD.includes(prev[prev.length - 1])) continue;
      if (isTitlecaseWord(prev) && isTitlecaseWord(words[i])) pairs.add(i);
    }
    return pairs;
  }

  function remainingFits(words, start, linesLeft, cap, tailTolerance) {
    if (start >= words.length) return true;
    if (linesLeft <= 0) return false;
    const budget = cap + tailTolerance;
    let used = 0;
    let cur = 0;
    for (let k = start; k < words.length; k++) {
      const wl = words[k].length;
      if (cur === 0) {
        cur = wl;
      } else if (cur + 1 + wl <= budget) {
        cur += 1 + wl;
      } else {
        used++;
        cur = wl;
        if (used >= linesLeft) return false;
      }
    }
    return used + (cur > 0 ? 1 : 0) <= linesLeft;
  }

  function wrapEn(text, options) {
    const cap = (options && options.cap) || 42;
    const maxLines = (options && options.maxLines) || 2;
    const tailTolerance = (options && options.tailTolerance != null) ? options.tailTolerance : 4;

    const trimmed = (text || "").trim();
    if (!trimmed) return { lines: [], hardCut: false };
    if (trimmed.length <= cap + tailTolerance) return { lines: [trimmed], hardCut: false };

    const words = trimmed.split(/\s+/);
    const lockedPairs = detectTitlecasePairs(words);
    const lines = [];
    let i = 0;
    let hardCut = false;

    while (i < words.length && lines.length < maxLines) {
      if (lines.length === maxLines - 1) {
        lines.push(words.slice(i).join(" "));
        i = words.length;
        break;
      }

      let bestJ = -1;
      let bestScore = -Infinity;
      let curLen = 0;
      let latestFittingJ = i + 1;

      for (let j = i; j < words.length; j++) {
        const wl = words[j].length;
        const newLen = curLen + (j > i ? 1 : 0) + wl;
        if (newLen > cap + tailTolerance) {
          if (j === i) {
            curLen = wl;
            latestFittingJ = i + 1;
          }
          break;
        }
        curLen = newLen;
        latestFittingJ = j + 1;

        if (j + 1 >= words.length) continue;

        const distance = Math.abs(curLen - cap);
        let score = 10;
        const lastCh = words[j].length ? words[j][words[j].length - 1] : "";
        if (EN_HARD.includes(lastCh)) score = 100;
        else if (EN_SOFT.includes(lastCh)) score = 70;

        const nxtClean = words[j + 1].replace(/[^\w]/g, "").toLowerCase();
        if (EN_CONNECTORS.has(nxtClean)) score = Math.max(score, 50);
        else if (EN_PREPOSITIONS.has(nxtClean)) score -= 40;  // v3: penalise prep-front

        if (lockedPairs.has(j + 1)) score -= 80;
        score -= distance * 2;

        const linesLeft = maxLines - lines.length - 1;
        if (!remainingFits(words, j + 1, linesLeft, cap, tailTolerance)) continue;

        if (score > bestScore) {
          bestScore = score;
          bestJ = j + 1;
        }
      }

      if (bestJ <= i) {
        bestJ = latestFittingJ;
        const linesLeft = maxLines - lines.length - 1;
        if (!remainingFits(words, bestJ, linesLeft, cap, tailTolerance)) hardCut = true;
      }

      lines.push(words.slice(i, bestJ).join(" "));
      i = bestJ;
    }

    if (i < words.length && lines.length > 0) {
      lines[lines.length - 1] = lines[lines.length - 1] + " " + words.slice(i).join(" ");
    }

    if (lines.some(l => l.length > cap + tailTolerance)) hardCut = true;
    return { lines, hardCut };
  }

  function resolveWrapConfig(fontConfig) {
    fontConfig = fontConfig || {};
    const standard = fontConfig.subtitle_standard;
    const basePreset = PRESETS[standard] || PRESETS[DEFAULT_PRESET];
    const zhCfg = Object.assign({}, basePreset.zh);
    const enCfg = Object.assign({}, basePreset.en);
    const explicit = fontConfig.line_wrap || {};
    const enabled = explicit.enabled != null ? explicit.enabled : true;
    // Explicit overrides apply to BOTH sub-configs (legacy single-cap compat)
    ["line_cap", "max_lines", "tail_tolerance"].forEach(key => {
      if (explicit[key] != null) {
        zhCfg[key] = explicit[key];
        enCfg[key] = explicit[key];
      }
    });
    return { enabled, zh: zhCfg, en: enCfg };
  }

  function wrapWithConfig(text, fontConfig) {
    const cfg = resolveWrapConfig(fontConfig);
    if (!cfg.enabled) {
      const trimmed = (text || "").trim();
      return { lines: trimmed ? [trimmed] : [], hardCut: false };
    }
    const sub = isZhText(text) ? cfg.zh : cfg.en;
    if (isZhText(text)) {
      return wrapZh(text, {
        cap: sub.line_cap,
        maxLines: sub.max_lines,
        tailTolerance: sub.tail_tolerance,
      });
    }
    return wrapEn(text, {
      cap: sub.line_cap,
      maxLines: sub.max_lines,
      tailTolerance: sub.tail_tolerance,
    });
  }

  // === v3.9 wrap_hybrid: soft target + hard cap + bottom-heavy + lock-aware ===
  // Mirror of backend/subtitle_wrap.py wrap_hybrid (Mod 5: F/B parity)

  function _scoreBreak(text, i) {
    // Score a break position. i is the line1 length (split between text[i-1] and text[i]).
    if (i < 1 || i > text.length) return 0;
    const ch = text[i - 1];
    if (HARD_BREAKS.includes(ch)) return 100;
    if (SOFT_BREAKS.includes(ch)) return 50;
    if (PAREN_CLOSE.includes(ch)) return 30;
    if (i < text.length && PAREN_OPEN.includes(text[i])) return 25;
    return 0;
  }

  function _findBestBreakInRange(text, lo, hi, locked) {
    // Return best break position in [lo, hi]. -1 if no scoring break found.
    let best = -1;
    let bestScore = -1;
    const n = text.length;
    hi = Math.min(hi, n);
    const start = Math.max(1, lo);
    for (let i = start; i <= hi; i++) {
      if (locked && i < locked.length && locked[i]) continue;
      let s = _scoreBreak(text, i);
      if (s > 0) {
        s += i; // tiebreaker: prefer later position
        if (s > bestScore) {
          bestScore = s;
          best = i;
        }
      }
    }
    return best;
  }

  // Python-compatible truthiness for `locked`:
  //   Python: bool([])=False, bool([False, False])=True, bool(None)=False
  //   JS equivalent: locked is non-null AND (if array, length > 0)
  function _lockedTruthy(locked) {
    if (locked == null) return false;
    if (Array.isArray(locked)) return locked.length > 0;
    return !!locked;
  }

  function wrapHybrid(text, opts) {
    opts = opts || {};
    let softCap = opts.soft_cap != null ? opts.soft_cap : 14;
    let hardCap = opts.hard_cap != null ? opts.hard_cap : 16;
    let maxLines = opts.max_lines != null ? opts.max_lines : 2;
    let tailTolerance = opts.tail_tolerance != null ? opts.tail_tolerance : 2;
    const locked = opts.locked || null;

    softCap = Math.max(1, softCap || 1);
    hardCap = Math.max(softCap, hardCap || softCap);
    maxLines = Math.max(1, maxLines || 1);
    tailTolerance = Math.max(0, tailTolerance || 0);

    const txt = (text || "").trim();
    if (!txt) {
      return {
        lines: [],
        hard_cut: false,
        soft_overflow: false,
        bottom_heavy_violation: false,
        lock_violated: false,
      };
    }

    const n = txt.length;

    // Single-line case 1: fits soft target
    if (n <= softCap + tailTolerance) {
      return {
        lines: [txt],
        hard_cut: false,
        soft_overflow: false,
        bottom_heavy_violation: false,
        lock_violated: false,
      };
    }

    // Single-line case 2: fits hard cap (band 2 — Netflix-compliant 1-line)
    if (n <= hardCap + tailTolerance) {
      return {
        lines: [txt],
        hard_cut: false,
        soft_overflow: true,
        bottom_heavy_violation: false,
        lock_violated: false,
      };
    }

    // Two-line wrap
    const lower = Math.max(1, n - hardCap);
    const upper = Math.min(hardCap, n - 1);

    if (lower > upper) {
      // Cannot fit in 2 lines without truncation (n > 2*hard_cap)
      const line1 = txt.slice(0, hardCap);
      const line2 = txt.slice(hardCap, hardCap * 2);
      return {
        lines: [line1, line2],
        hard_cut: true,
        soft_overflow: true,
        bottom_heavy_violation: line1.length > line2.length,
        lock_violated: false,
      };
    }

    const bhUpper = Math.min(upper, Math.floor(n / 2) + tailTolerance);

    // Pass 1: bottom-heavy
    if (bhUpper >= lower) {
      const best = _findBestBreakInRange(txt, lower, bhUpper, locked);
      if (best > 0) {
        const line1 = txt.slice(0, best).replace(/\s+$/, "");
        const line2 = txt.slice(best).replace(/^\s+/, "");
        return {
          lines: [line1, line2],
          hard_cut: false,
          soft_overflow: line1.length > softCap || line2.length > softCap,
          bottom_heavy_violation: line1.length > line2.length,
          lock_violated: false,
        };
      }
    }

    // Pass 2: full hard-cap range
    const best2 = _findBestBreakInRange(txt, lower, upper, locked);
    if (best2 > 0) {
      const line1 = txt.slice(0, best2).replace(/\s+$/, "");
      const line2 = txt.slice(best2).replace(/^\s+/, "");
      return {
        lines: [line1, line2],
        hard_cut: false,
        soft_overflow: line1.length > softCap || line2.length > softCap,
        bottom_heavy_violation: line1.length > line2.length,
        lock_violated: false,
      };
    }

    // Pass 3: any unlocked scoring break in [1, n-1] — only fires when locked truthy
    if (_lockedTruthy(locked)) {
      const anyUnlocked = _findBestBreakInRange(txt, 1, n - 1, locked);
      if (anyUnlocked > 0) {
        const line1 = txt.slice(0, anyUnlocked).replace(/\s+$/, "");
        const line2 = txt.slice(anyUnlocked).replace(/^\s+/, "");
        return {
          lines: [line1, line2],
          hard_cut: false,
          soft_overflow: true,
          bottom_heavy_violation: line1.length > line2.length,
          lock_violated: false,
        };
      }
    }

    // Pass 4: forced hard-cut
    const cut = Math.min(hardCap, Math.floor(n / 2));
    const line1 = txt.slice(0, cut);
    let line2 = txt.slice(cut);
    if (line2.length > hardCap) line2 = line2.slice(0, hardCap);
    return {
      lines: [line1, line2],
      hard_cut: true,
      soft_overflow: line1.length > softCap || line2.length > softCap,
      bottom_heavy_violation: line1.length > line2.length,
      lock_violated: _lockedTruthy(locked),
    };
  }

  window.SubtitleWrap = { wrapZh, wrapEn, wrapHybrid, wrapWithConfig, resolveWrapConfig, isZhText, PRESETS };
})();
